### PR TITLE
[quant][core][gpu][feature] Added support for float->quantized cuda tensor copying

### PR DESCRIPTION
**Pull request by dzdang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #76184
* __->__ #76177

Summary:
Previously, support for copying a fp tensor to a quantized tensor was
limited to CPU tensors. This PR extends the support to GPU tensors.
A corresponding test was added to test_qtensor_float_assignment for cuda
tensors

Test plan:
```
python test/test_quantization.py -k test_qtensor_float_assignment
```

Differential Revision: [D35817832](https://our.internmc.facebook.com/intern/diff/D35817832)

### CODE REVIEW

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35817832).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35817832).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35817832).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35817832).

**Comment by dzdang** [comment]
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35817832).

File: test/quantization/core/test_quantized_tensor.py [hunk 1]
```diff
@@ -354,25 +354,33 @@ def test_qtensor_float_assignment(self):
...
-            # Copying from a float Tensor
-            qr[:] = x
-            self.assertEqual(qr.item(), 15)
-
-            dtype_msg = str(dtype) + ", "
-            self.assertEqual(' '.join(str(qr).split()),
-                             "tensor([15.], size=(1,), dtype=" + dtype_msg +
-                             "quantization_scheme=torch.per_tensor_affine, " +
-                             "scale=1.0, zero_point=2)")
+        devices = ["cpu", "cuda"] if torch.cuda.is_available() else ["cpu"]
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_tensor.py:357]: wondering if we can factor this out as an individual helper function in this file or in common_quantization.py

File: test/quantization/core/test_quantized_tensor.py [hunk 2]
```diff
@@ -354,25 +354,33 @@ def test_qtensor_float_assignment(self):
...
+                qr[0] = torch.Tensor([11.3]).to(device=device)  # float assignment
+                self.assertEqual(qr.item(), 11)
+                x = torch.ones(1, dtype=torch.float).to(device=device) * 15.3
+                # Copying from a float Tensor
+                qr[:] = x
+                self.assertEqual(qr.item(), 15)
+
+                dtype_msg = str(dtype) + ", "
+                if device == "cuda":
+                    self.assertEqual(' '.join(str(qr).split()),
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_tensor.py:375]: I'm a bit confused, why are we doing this check?

**Review by jerryzh168 — APPROVED** [review]

**Pull request closed without merging**